math/mathcore/inc/Fit/EvaluateLogL.hxx

```diff
@@ -0,0 +1,569 @@
+
+   @ingroup FitMain
+*/
+namespace FitUtil {
+
+  // internal class defining
+  template <class T>
+  class LikelihoodAux {
+  public:
+     LikelihoodAux(T logv = {}, T w = {}, T w2 = {}) : logvalue(logv), weight(w), weight2(w2) {}
+
+     LikelihoodAux operator+(const LikelihoodAux &l) const
+     {
+        return LikelihoodAux<T>(logvalue + l.logvalue, weight + l.weight, weight2 + l.weight2);
+     }
+
+     LikelihoodAux &operator+=(const LikelihoodAux &l)
+     {
+        logvalue += l.logvalue;
+        weight += l.weight;
+        weight2 += l.weight2;
+        return *this;
+     }
+
+     T logvalue;
+     T weight;
+     T weight2;
+  };
+
+  template <>
+  class LikelihoodAux<double> {
+  public:
+     LikelihoodAux(double logv = 0.0, double w = 0.0, double w2 = 0.0) : logvalue(logv), weight(w), weight2(w2){};
+
+     LikelihoodAux operator+(const LikelihoodAux &l) const
+     {
+        return LikelihoodAux<double>(logvalue + l.logvalue, weight + l.weight, weight2 + l.weight2);
+     }
+
+     LikelihoodAux &operator+=(const LikelihoodAux &l)
+     {
+        logvalue += l.logvalue;
+        weight += l.weight;
+        weight2 += l.weight2;
+        return *this;
+     }
+
+     double logvalue;
+     double weight;
+     double weight2;
+  };
+
+
+  /**
+      evaluate the LogL given a model function and the data at the point x.
+      return also nPoints as the effective number of used points in the LogL evaluation
+  */
+  double EvaluateLogL(const IModelFunction &func, const UnBinData &data, const double *p, int iWeight, bool extended,
+                      unsigned int &nPoints, ROOT::Fit::ExecutionPolicy executionPolicy, unsigned nChunks = 0);
+
+  /**
+      evaluate the LogL gradient given a model function and the data at the point x.
+      return also nPoints as the effective number of used points in the LogL evaluation
+  */
+  void EvaluateLogLGradient(const IModelFunction &func, const UnBinData &data, const double *x, double *grad,
+                            unsigned int &nPoints,
+                            ROOT::Fit::ExecutionPolicy executionPolicy = ROOT::Fit::ExecutionPolicy::kSerial,
+                            unsigned nChunks = 0);
+
+    /**
+      evaluate the pdf contribution to the LogL given a model function and the BinPoint data.
+      If the pointer g is not null evaluate also the gradient of the pdf.
+      If the function provides parameter derivatives they are used otherwise a simple derivative calculation
+      is used
+  */
+  double EvaluatePdf(const IModelFunction &func, const UnBinData &data, const double *x, unsigned int ipoint, double *g = 0);
+  
+  #ifdef R__HAS_VECCORE
+     template <class NotCompileIfScalarBackend = std::enable_if<!(std::is_same<double, ROOT::Double_v>::value)>>
+     double EvaluatePdf(const IModelFunctionTempl<ROOT::Double_v> &func, const UnBinData &data, const double *p, unsigned int i, double *) {
```

Comment:
From conversations with Axel, I think we should avoid at all costs to put explicitly `ROOT::Double_v` in any interface used in ROOT.

Reply:
I will see to remove it, but If it's there, it's because at the time we thought it was unavoidable.